#!/bin/bash -l

#SBATCH --account=cdebes
#SBATCH --error=/data/public/cdebes/star_hsapiens_blood.log
#SBATCH --job-name=hs
#SBATCH --partition=all
#SBATCH --ntasks=16

module load STAR-2.6.0a
#module unload R-3.2.3
#module load R-3.3.1

path=$1
filea=$2

#filea1=$filea\_1.fq.gz
#filea2=$filea\_2.fq.gz

#filea1=$filea\_1.fastq.gz
#filea2=$filea\_2.fastq.gz

filea1=$filea\_1_sequence.fq.gz
filea2=$filea\_2_sequence.fq.gz

#echo $path$filea1
#echo $path$filea2

#filea1=$filea\_R1_001_f.fastq.gz
#filea2=$filea\_R2_001_f.fastq.gz

#filea1=$filea\.fastq
#filea1=$filea\.fq.gz
#filea1=$filea\.fastq.gz

#filea1=$filea\_1.fastq
#filea2=$filea\_2.fastq

#mkdir $path/$filea/
#java -jar /data/public/cdebes/soft/Trimmomatic-0.33/trimmomatic-0.33.jar PE $path$filea1 $path$filea2 $path$filea/output_forward_paired.fq.gz $path$filea/output_forward_unpaired.fq.gz $path$filea/output_reverse_paired.fq.gz $path$filea/output_reverse_unpaired.fq.gz ILLUMINACLIP:../../soft/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:35

#java -jar /data/public/cdebes/soft/Trimmomatic-0.33/trimmomatic-0.33.jar PE $path$filea1 $path$filea2 $path$filea/output_forward_paired.fq.gz $path$filea/output_forward_unpaired.fq.gz $path$filea/output_reverse_paired.fq.gz $path$filea/output_reverse_unpaired.fq.gz LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:65

#java -jar /data/public/cdebes/soft/Trimmomatic-0.33/trimmomatic-0.33.jar SE $path$filea1 $path$filea/output_forward_paired.fq.gz ILLUMINACLIP:../../soft/Trimmomatic-0.33/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:45

#STAR --runThreadN 54 --genomeDir /data/public/mgarmha1/Assemblies/Mouse/Starindex_75 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix $path/$filea/ --readFilesCommand zcat --outSAMtype BAM Unsorted
#--outWigType bedGraph --outWigNorm RPM

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 32 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_prog/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 

#GRO-seq
#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 24 --readFilesIn $path$filea1  --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens/$filea --outSAMtype BAM Unsorted SortedByCoordinate  --outSAMstrandField intronMotif  --readFilesCommand zcat

#microglia
#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outSAMstrandField intronMotif

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 32 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_mnase/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigStrand Unstranded

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_new2/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15  --chimSegmentMin 15

#TO DO
#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf  --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz  --outFileNamePrefix /data/public/cdebes/workspace/scripts/fuchs_labeling/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --sjdbGTFchrPrefix chr --readFilesCommand zcat --outFilterMatchNmin 20 --outFilterScoreMinOverLread 0 --outFilterMatchNminOverLread 0 --outFilterMismatchNmax 35 --outWigType bedGraph --outWigNorm None

#STAR --genomeDir /data/public/cdebes/workspace/genomes/dmelanogaster/ --runThreadN 32 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/dmelanogaster_mut/$filea --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/dmelanogaster/Drosophila_melanogaster.BDGP6.90.gtf 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/dmelanogaster/ --runThreadN 32 --readFilesIn $path$filea1  --outFileNamePrefix /data/public/cdebes/workspace/scripts/dmelanogaster_merged/$filea --outSAMtype BAM SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/dmelanogaster/Drosophila_melanogaster.BDGP6.90.gtf --readFilesCommand zcat

#STAR --genomeDir /data/public/cdebes/workspace/genomes/dmelanogaster/ --runThreadN 32 --readFilesIn $path$filea/output_forward_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/dmelanogaster_merged/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/dmelanogaster/Drosophila_melanogaster.BDGP6.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outSAMstrandField intronMotif  --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimSegmentMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None 
#echo $path$filea1

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/dmelanogaster/ --runThreadN 16 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/dmelanogaster_mut/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/dmelanogaster/Drosophila_melanogaster.BDGP6.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate  --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 4 --alignSJDBoverhangMin 4 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --outWigType bedGraph --outWigNorm None --readFilesCommand zcat 

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/rnorvegicus/ --runThreadN 16 --readFilesIn $path$filea1  --outFileNamePrefix /data/public/cdebes/workspace/scripts/rnorvegicus/$filea --outSAMtype BAM Unsorted SortedByCoordinate  --sjdbGTFfile /data/public/cdebes/workspace/genomes/rnorvegicus/Rattus_norvegicus.Rnor_6.0.85.gtf --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None --readFilesCommand zcat 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus_ensembl/ --runThreadN 50 --readFilesIn $path$filea1  $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_hypoxia/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.87.gtf --sjdbGTFchrPrefix chr --readFilesCommand zcat --outWigType bedGraph  --outWigNorm RPM 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 32 --readFilesIn $path$filea1 $path$filea2 --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_kidney/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 32 --readFilesIn $path$filea1 $path$filea2 --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_nascent/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 5 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 16 --readFilesIn $path$filea1 --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_4udrb_xue/$filea --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif

#STAR --genomeDir /data/public/cdebes/workspace/genomes/hsapiens/ --runThreadN 32 --readFilesIn  $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/hsapiens_sampling/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/hsapiens/Homo_sapiens.GRCh38.85.gtf --outSAMtype BAM Unsorted SortedByCoordinate  --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 32 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_hypocampus/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat  --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 2 --alignSJDBoverhangMin 2 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 32 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz  --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_blood/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat  --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None

/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 16 --readFilesIn  $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 16 --readFilesIn  $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_hypocampus/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMtype BAM Unsorted SortedByCoordinate --readFilesCommand zcat  --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 16 --readFilesIn $path$filea1 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_hypothalamus/$filea --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf  --outSAMtype BAM Unsorted SortedByCoordinate --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --outWigType bedGraph --outWigNorm None

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 32 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /cellnet/SyBACol/markus/sofsdav.uni-koeln.de/nephrolab/Labor/DataExchange/for_Cedric/$filea/ --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMtype BAM Unsorted --outSAMstrandField intronMotif --readFilesCommand zcat

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 32 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_wat/$filea --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outSAMstrandField intronMotif --outSAMtype BAM SortedByCoordinate --outWigType bedGraph --outWigNorm None 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_whiteadipose_polya/$filea --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.87.gtf --sjdbGTFchrPrefix chr --outWigType bedGraph --outWigNorm None --outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif

#STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 24 --readFilesIn  $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_wat/$filea --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.87.gtf --outSAMtype BAM SortedByCoordinate Unsorted --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/fromElifeLys/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outWigType bedGraph --outWigNorm None --outSAMtype BAM Unsorted SortedByCoordinate --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/mmusculus/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/mmusculus_blood/$filea --sjdbGTFfile /data/public/cdebes/workspace/genomes/mmusculus/Mus_musculus.GRCm38.90.gtf --outWigType bedGraph --outWigNorm None --outSAMtype BAM Unsorted SortedByCoordinate --outSAMstrandField intronMotif --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 --readFilesCommand zcat

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/celegans_mut/$filea --outSAMstrandField intronMotif --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf  --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15  --outWigType bedGraph --outWigNorm None 

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/ssl1/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile  /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf --readFilesCommand zcat  --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 32 --readFilesIn $path$filea1/output_forward_paired.fq.gz $path$filea2/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/celegans/$filea --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/celegans_will_mair/$filea --outSAMtype BAM Unsorted SortedByCoordinate  --outWigType bedGraph --outWigNorm None --outSAMstrandField intronMotif --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf 

#chimeric on
#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/ssl1/$filea --outSAMtype BAM Unsorted SortedByCoordinate --outWigType bedGraph --outWigNorm None --readFilesCommand zcat --outSAMstrandField intronMotif --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/rnp-6/$filea --outSAMtype BAM Unsorted SortedByCoordinate --outWigType bedGraph --outWigNorm None --readFilesCommand zcat --outSAMstrandField intronMotif --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/CHK1/$filea --outSAMtype BAM Unsorted SortedByCoordinate  --outWigType bedGraph --outWigNorm None --readFilesCommand zcat --outSAMstrandField intronMotif --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf  --chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15

#STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 24 --readFilesIn $path$filea/output_forward_paired.fq.gz $path$filea/output_reverse_paired.fq.gz --outFileNamePrefix /data/public/cdebes/workspace/scripts/ssl1_chip_seq/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.87.gtf --readFilesCommand zcat  --outWigType bedGraph --outWigNorm None --outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif --outWigStrand Unstranded

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 16 --readFilesIn $path$filea1 $path$filea2 --outFileNamePrefix /data/public/cdebes/workspace/scripts/celegans_mut/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf --outWigType bedGraph --outWigNorm None --outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif --readFilesCommand zcat --alignSJDBoverhangMin 5 --chimSegmentMin 15 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 

#/data/programs/STAR-2.6.0a-haswell/bin/Linux_x86_64/STAR --genomeDir /data/public/cdebes/workspace/genomes/celegans/ --runThreadN 16 --readFilesIn $path$filea1 --outFileNamePrefix /data/public/cdebes/workspace/scripts/celegans_polya/$filea --outSAMtype BAM Unsorted SortedByCoordinate --sjdbGTFfile /data/public/cdebes/workspace/genomes/celegans/Caenorhabditis_elegans.WBcel235.90.gtf --outWigType bedGraph --outWigNorm None --outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif --readFilesCommand zcat --alignSJDBoverhangMin 5

#Optimized chimeric RNA/fusion transcript
#--outSJfilterOverhangMin 10 10 10 10 --outSJfilterCountUniqueMin 1 1 1 1 --outSJfilterIntronMaxVsReadN 400000 400000 400000 400000 --outSJfilterDistToOtherSJmin 0 0 0 0 --outSAMstrandField intronMotif --readFilesCommand zcat  --twopassMode Basic --outReadsUnmapped None --chimSegmentMin 12 --chimJunctionOverhangMin 12 --alignSJDBoverhangMin 10--alignMatesGapMax 100000 --alignIntronMax 100000 --chimSegmentReadGapMax 3 --alignSJstitchMismatchNmax 5 -1 5 5
#Optimized circRNA
#-chimSegmentMin 15 --outSJfilterOverhangMin 15 15 15 15 --alignSJoverhangMin 15 --alignSJDBoverhangMin 15 --seedSearchStartLmax 30 --outFilterMultimapNmax 10 --outFilterScoreMin 1 --outFilterMatchNmin 1 --outFilterMismatchNmax 2 --chimScoreMin 15 --chimScoreSeparation 10 --chimJunctionOverhangMin 15 
